aten/src/ATen/native/ForeachUtils.h

```diff
@@ -173,5 +180,140 @@ bool can_use_fast_route(TensorList tensors1, TensorList tensors2, bool does_op_p
   return can_use_fast_route({tensors1, tensors2}, {}, does_op_promote_integer_inputs_to_float);
 }
 
+using nested_tensorvec_t = std::vector<std::vector<at::Tensor>>;
+using scalartype_nested_tensorvec_map_t = std::unordered_map<at::ScalarType, nested_tensorvec_t>;
+using device_scalartype_nested_tensorvec_map_t = std::unordered_map<at::Device, scalartype_nested_tensorvec_map_t>;
+using nested_tensorvec_t = std::vector<std::vector<at::Tensor>>;
+using scalartype_nested_tensorvec_map_t = std::unordered_map<at::ScalarType, nested_tensorvec_t>;
+using device_scalartype_nested_tensorvec_map_t = std::unordered_map<at::Device, scalartype_nested_tensorvec_map_t>;
+using DeviceDtypeKey = std::pair<at::Device, at::ScalarType>;
+using IndicesT = std::vector<int>;
+using TensorsAndIndicesT = std::pair<nested_tensorvec_t, IndicesT>;
+using FlatMap = std::unordered_map<DeviceDtypeKey, TensorsAndIndicesT, ParamsHash<DeviceDtypeKey>>;
+
+FlatMap group_tensors_by_first_tensors_device_and_dtype(const std::vector<std::vector<Tensor>>& nested_tensorlist, const bool with_indices) {
+  FlatMap grouped_tensors_with_indices;
+
+  TORCH_CHECK_GT(nested_tensorlist.size(), 0);
+  TORCH_CHECK_GT(nested_tensorlist[0].size(), 0);
+  const auto num_lists = nested_tensorlist.size();
+  const auto num_tensors = nested_tensorlist[0].size();
+
+  TORCH_CHECK(std::all_of(nested_tensorlist.cbegin(), nested_tensorlist.cend(),
+    [&](const auto& tensorlist) -> bool {
+      // note(crcrpar): Allow empty tensorlists following
+      // ref: https://github.com/pytorch/pytorch/blob/85885301fd3c6adb8b9dc3cf7afadf6945566684/torch/utils/_foreach_utils.py#L21-L24
+      return tensorlist.size() == num_tensors || tensorlist.size() == 0;
+    }));
+
+  for (const auto& tensor_index : c10::irange(num_tensors)) {
+    const auto t = nested_tensorlist[0][tensor_index];
+    const DeviceDtypeKey key = {t.device(), t.scalar_type()};
+    if (!grouped_tensors_with_indices.count(key)) {
+      grouped_tensors_with_indices.insert(
+        {
+          key,
+          TensorsAndIndicesT{
+            [&]() -> nested_tensorvec_t {
+              nested_tensorvec_t nested_tensorvec;
+              nested_tensorvec.reserve(num_lists);
+             for (const auto& i : c10::irange(num_lists)) {
+                std::vector<Tensor> tensors;
+                if (!nested_tensorlist[i].empty()) {
+                  tensors.reserve(num_tensors);
+                }
+                nested_tensorvec.emplace_back(tensors);
+              }
+              return nested_tensorvec;
+            }(),
+            [&]() -> IndicesT {
+              if (!with_indices) {
+                return {};
+              } else {
+                IndicesT indices;
+                indices.reserve(num_tensors);
+                return indices;
+              }
+            }()
+          }
+        }
+      );
+    }
```

Comment:
should this logic just be moved before the for loop so it doesn't keep checking with count?